cumulusci/cli/cci.py

```diff
@@ -872,13 +871,12 @@ def flow_run(config, flow_name, org, delete_org, debug, o, skip, no_prompt):
     try:
         flow = flow_class(config.project_config, flow_config,
                           org_config, options, skip, name=flow_name)
-
         flow()
     except (TaskRequiresSalesforceOrg, TaskOptionsError) as e:
         exception = click.UsageError(e.message)
         handle_exception_debug(config, debug, throw_exception=exception)
     except (ApexTestException, BrowserTestFailure, MetadataComponentFailure,
-            MetadataApiError, ScratchOrgException) as e:
+            MetadataApiError, RobotTestFailure, ScratchOrgException) as e:
```

Comment:
Rather than hardcoding a big list of exceptions here, how about if we define CumulusCIExpectedFailure and make the exceptions listed here subclass it? Then we can just catch CumulusCIExpectedFailure here, which makes it simpler for CumulusCI extensions to define their own exceptions (like if we want to refactor the robot support into a separate package)

For the usage errors, maybe TaskRequiresSalesforceOrg and TaskOptionsError should actually be subclasses of click.UsageError

Reply:
chris always had dreams of making our exceptions more of a hierarchy, this is probably a good idea.

Reply:
I'm not sure what we'd do if we want to handle an exception that we don't define ourselves as an expected failure, though.

Reply:
I changed this to use the new exception classes CumulusCIFailure and CumulusCIUsageError.  I also added a new section to exception handling in task_run and flow_run which currently only has ScratchOrgException but could be used to handle other exceptions from outside of CumulusCI.  ScratchOrgException is a special case where it seems to make sense to allow the implementation to decide how they want to handle the exception.

I didn't have CumulusCIUsageError subclass click.UsageError since click is only used in the CLI and not in other implementations (such as MetaCI).  In MetaCI, we'll likely want to have CumulusCIUsageError create an Error status on a build while all exceptions in CumulusCIFailure will create a Fail status.